doozer/doozerlib/cli/scan_sources.py

```diff
@@ -54,6 +69,162 @@ def run(self):
         # We have our information. Now build and print the output report
         self.generate_report()
 
+    def _try_reconciliation(self, metadata: Metadata, repo_name: str, pub_branch_name: str, priv_branch_name: str):
+        reconciled = False
+
+        # Attempt a fast-forward merge
+        rc, _, _ = exectools.cmd_gather(cmd=['git', 'pull', '--ff-only', 'public_upstream', pub_branch_name])
+        if not rc:
+            # fast-forward succeeded, will push to openshift-priv
+            self.runtime.logger.info('Fast-forwarded %s from public_upstream/%s', metadata.name, pub_branch_name)
+            reconciled = True
+
+        else:
+            # fast-forward failed, trying a merge commit
+            rc, _, _ = exectools.cmd_gather(cmd=['git', 'merge', f'public_upstream/{pub_branch_name}',
+                                                 '-m', f'Reconciled {repo_name} with public upstream'],
+                                            log_stderr=True,
+                                            log_stdout=True)
+            if not rc:
+                # merge succeeded, will push to openshift-priv
+                reconciled = True
+                self.runtime.logger.info('Merged public_upstream/%s into %s', priv_branch_name, metadata.name)
+
+        if not reconciled:
+            # Could not rebase from public upstream: need manual reconciliation. Log a warning and return
+            self.runtime.logger.warning('failed rebasing %s from public upstream: will need manual reconciliation',
+                                        metadata.name)
+            self.issues.append({'name': metadata.distgit_key,
+                                'issue': 'Could not rebase into -priv as it needs manual reconciliation'})
+            return
+
+        if self.dry_run:
+            self.runtime.logger.info('Would have tried reconciliation for %s/%s', repo_name, priv_branch_name)
+            return
+
+        # Try to push to openshift-priv
+        try:
+            exectools.cmd_assert(
+                cmd=['git', 'push', 'origin', priv_branch_name],
+                retries=3)
+
+            # Successfully pushed: mark repo as changed
+            message = f'{metadata.name} reconciled with public openshift/{repo_name}'
+            rebuild_hint = RebuildHint(RebuildHintCode.REBASE_INTO_PRIV, message)
+            self.runtime.logger.info(message)
+
+            if metadata.meta_type == 'rpm':
+                self.add_assessment_reason(metadata, rebuild_hint)
+                self.changing_rpm_metas.add(metadata)
+
+            elif metadata.meta_type == 'image':
+                self.add_image_meta_change(metadata, rebuild_hint)
```

Comment:
After today's excursion with https://github.com/openshift-eng/art-tools/pull/112 , I'm not sure if we should add this hint here. Because `scan_for_upstream_changes` which gets run after this will detect that there is no build for the commit, it also checks if image/rpm isn't `mode:disabled ` (which I think we can do in several ways). There's also `needs_rebuild` and `RebuildHintCode.DELAYING_NEXT_ATTEMPT` which seem like special cases where we don't want to rebuild.

Reply:
This reminds me maybe we shouldn't sync disabled images/rpms..

Reply:
Turns out doozer will exclude disabled images by default (they do not appear in the image map)

Reply:
skipping rebase for disabled rpms/images: 0fb153af57bf22d1c6e1fdd4de6ec9a3172a0ccd

Reply:
Regarding the rebuild hint, good catch. Removed in 6c659302ff725c6744b6a6cb1fa97fb44d3f43b6